allennlp/semparse/worlds/atis_world.py

```diff
@@ -27,34 +51,72 @@ def __init__(self, utterances: List[str], tokenizer=None) -> None:
         self.utterances: List[str] = utterances
         self.tokenizer = tokenizer if tokenizer else WordTokenizer()
         self.tokenized_utterances = [self.tokenizer.tokenize(utterance) for utterance in self.utterances]
-        self.valid_actions: Dict[str, List[str]] = self.init_all_valid_actions()
+        valid_actions, linking_scores = self.init_all_valid_actions()
+        self.valid_actions: Dict[str, List[str]] = valid_actions
+
+        # This has shape (number_entities, number_utterance_tokens).
+        self.linking_scores: numpy.ndarray = linking_scores
         self.grammar_str: str = self.get_grammar_str()
         self.grammar_with_context: Grammar = Grammar(self.grammar_str)
 
     def get_valid_actions(self) -> Dict[str, List[str]]:
         return self.valid_actions
 
-    def init_all_valid_actions(self) -> Dict[str, List[str]]:
+    def init_all_valid_actions(self) -> Tuple[Dict[str, List[str]], numpy.ndarray]:
         """
-        We initialize the world's valid actions with that of the context. This means that the strings
-        and numbers that were valid earlier in the interaction are also valid. We then add new valid strings
-        and numbers from the current utterance.
+        We initialize the valid actions with the global actions. We then iterate through the
+        utterances up to and including the current utterance and add the valid strings.
         """
+
         valid_actions = deepcopy(self.sql_table_context.valid_actions)
-        for string in self.get_strings_from_utterance():
+        linking_scores = []
+        current_tokenized_utterance = [] if not self.tokenized_utterances \
+                else self.tokenized_utterances[-1]
+
+        strings: Set[str] = set()
+        for tokenized_utterance in self.tokenized_utterances:
+            string_linking_dict = get_strings_from_utterance(tokenized_utterance)
+            strings = strings.union(set(string_linking_dict.keys()))
+
+        strings_list: List[str] = sorted(strings, reverse=True)
```

Comment:
Why `reverse=True`?

Reply:
I'm setting ``reverse=True`` for consistency with the grammar, since strings are sorted in reverse there. We want this because the parser is greedy (if we have a rule that has multiple options for the right hand side, the first one that succeeds is the one that is used). For example, if we have `1400` in the query and both `1400` and `1` are valid numbers, then we want to try to match it with `1400` first. Otherwise, it will consume `1` and then fail to match anything with `400`.

Reply:
Ah, you're catering to `parsimonious`?  That's fine, but giving a note explaining that would be helpful.